local Players = game:GetService("Players")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
require(ReplicatedStorage.Shared.network)
local Admin = require(script.Admin.Admin)
local AdminRights = require(script.Admin.AdminRights)
local StockState = require(script.Modules.Shops.StockState)
local TransactionHandler = require(script.Modules.Shops.TransactionHandler)
local CurrencyService = require(script.Services.CurrencyService)
local DataStore = require(script.Services.DataStore)
local DialogueServer = require(script.Services.DialogueServer)
local EntityService = require(script.Services.EntityService)
require(script.Services.DialogueServer)
require(script.Services.InventoryServer)
local IslandService = require(script.Services.IslandService)
local PlayerManager = require(script.Services.PlayerManager)
local QuestService = require(script.Services.QuestService)
local ShopService = require(script.Services.ShopService)
local TreeService = require(script.Services.TreeService)

Players.PlayerAdded:Connect(function(player: Player)
	player:LoadCharacter()

	local character = player.Character or player.CharacterAdded:Wait()
	character:WaitForChild("HumanoidRootPart"):PivotTo(CFrame.new(Vector3.new(100, 0, 0)))
	task.wait(0.2)
	PlayerManager.PlayerJoined(player)
	DataStore.LoadPlayerData(player)
	IslandService.PlayerJoined(player)
	EntityService.PlayerJoined(player)
	TreeService.PlayerJoined(player)
	CurrencyService.PlayerJoined(player)
	QuestService.PlayerJoined(player)
	StockState.Init(player)
	ShopService.Init(player)

	if AdminRights.HasRights(player) then
		Admin.Init(player)
	end
end)

EntityService.Start()
TransactionHandler.Init()
DialogueServer.Init()

Players.PlayerRemoving:Connect(function(player: Player)
	IslandService.PlayerLeaving(player)
	EntityService.PlayerLeaving(player)
end)
