local Players = game:GetService("Players")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local TreeManager = require(script.Modules.TreeManager)
local InventoryManager = require(script.Modules.InventoryManager)
require(ReplicatedStorage.Shared.network)
local IslandManager = require(script.Modules.IslandManager)

Players.PlayerAdded:Connect(function(player: Player)
	IslandManager.PlayerJoined(player)
	InventoryManager.PlayerJoined(player)
	TreeManager.PlayerJoined(player)
end)

Players.PlayerRemoving:Connect(function(player: Player)
	IslandManager.PlayerLeaving(player)
	InventoryManager.PlayerLeaving(player)
end)
